### PR TITLE
Add Databag Messenger

### PR DESCRIPTION
**Pull request by balzack**
# App Submission

### App name
Databag

### 256x256 SVG icon
https://s3.us-west-2.amazonaws.com/org.coredb.react/app_icon.svg

### Gallery images
https://s3.us-west-2.amazonaws.com/org.coredb.react/mobile.png
https://s3.us-west-2.amazonaws.com/org.coredb.react/browser.png
https://s3.us-west-2.amazonaws.com/org.coredb.react/connect.png

### I have tested my app on:
- [-] [Umbrel dev environment](https://github.com/getumbrel/umbrel-dev)
- [-] [Umbrel OS on a Raspberry Pi 4](https://github.com/getumbrel/umbrel-os)
- [-] [Custom Umbrel install on Linux](https://github.com/getumbrel/umbrel#-installation)

### CODE REVIEW

**Comment by balzack** [comment]
I was getting port conflicts during my testing, so I switched to port 7007, and that worked.
The browser app works on plain http, but the mobile app requires https; this was tested and verified using nginx proxy manager forwarding to a RPI4 umbrel install.

**Comment by nmfretz** [comment]
Thanks for submitting Databag Messenger @balzack! This looks great. I will take a look at this PR in detail and review next week.

**Comment by balzack** [comment]
Awesome...thank you!

File: databag/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+version: "3.7"
+
+services:
+
+  app_proxy:
+    environment:
+      APP_HOST: databag_web_1
+      APP_PORT: 7777
```
Comment (IMPranshu) [inline databag/docker-compose.yml:8]:
```suggestion
      APP_PORT: 7000
```
as it is the port used inside the container exposed by docker image.

File: databag/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+      
+  web:
+    image: balzack/databag:latest
+    restart: on-failure
+    stop_grace_period: 1m
+    ports:
```
Comment (IMPranshu) [inline databag/docker-compose.yml:14]:
```suggestion
```
ports are not explicitly required as we are using `app_proxy` which handles the port mapping.

File: databag/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+      - "7007:7000"
```
Comment (IMPranshu) [inline databag/docker-compose.yml:15]:
```suggestion
```
You were getting port conflicts error because of this.

**Review by IMPranshu — CHANGES_REQUESTED** [review]
These changes will take care of the port conflict issue when starting up the app. Previously the app wouldn't have been installed from the Umbrel app store due to the error.

File: databag/docker-compose.yml (hunk 2 above)
Comment (balzack) [reply databag/docker-compose.yml:14]: @IMPranshu suggestion applied, thank you!

Comment (IMPranshu) [reply databag/docker-compose.yml]: Did you try running the app once after applying the changes? Do try once. Will help in opening the next app submission PR.

Comment (balzack) [reply databag/docker-compose.yml]: I am traveling for the next week; I will try when I return.

**Comment by balzack** [comment]
@IMPranshu these changes don't seem to work for me. 

I am able to access the webapp with the URL http://<IP>:7077, but this webapp needs to be accessible via https.

Before I was able to access the webapp through a reverse proxy, but now it seems to keep redirecting me to an inaccessible login.

Is the merge request not valid the way it was initially?

**Comment by IMPranshu** [comment]
Can you confirm what is the internal port for this image? 

Initially, the app wouldn't have been installed as it would have thrown an error of port already allocated as happened in my case, so technically not a valid PR but @nmfretz would know better as he is from the umbrel team.

**Comment by balzack** [comment]
@IMPranshu the internal port of the image is 7000.

It is possible that things are setup correctly, but I do not know how to verify. When I try and find the installed test app through the umbrel UI, I just see an empty placeholder in the appstore. I am not sure if this is expected considering the test install.

I am able to connect to my webapp through local IP and port, but when I try and connect externally through a reverse proxy, I am redirected to the umbrel login at port 2000. That port is not accessible externally.

**Comment by IMPranshu** [comment]
Reverse proxying to Umbrel is a bit tricky. Have you tried reverse proxying to any other official Umbrel apps? Is that working?

**Comment by balzack** [comment]
I wasnt able to get reverse proxy with other apps either, for that reason I think it might be configured correctly. 

Is there a guide to getting reverse proxy working with umbrel apps?

**Comment by IMPranshu** [comment]
https://jorijn.com/installing-nginx-reverse-proxy-with-ssl-certificate-umbrel-btcpayserver/

It's a blog that might help in this case.

**Comment by balzack** [comment]
I have reverse proxy working in general, just not when it comes to umbrel's login redirection to port 2000. This blog post doesn't mention that. I'll look around for other guides. Thanks.

**Comment by nmfretz** [comment]
Apologies for the delay in reviewing @balzack. I have had my head down on other areas of the code-base. This is still on my radar.

**Comment by balzack** [comment]
@nmfretz, no worries. I am also working on other areas of my project.

It might be in a good state as is after the changes suggested by @IMPranshu, but I am unable to test on my Pi. My home network has a reverse proxy forwarding to my various systems and services, and I can't seem to configure that with umbrel's authentication redirection.

**Comment by nmfretz** [comment]
Hey @balzack, sorry again for the delay in reviewing on my end. I went through this today and pushed some commits to get it working on Umbrel. Everything is working well! Here are the changes I made:
- https://github.com/getumbrel/umbrel-apps/pull/649/commits/35a7a9b4fd576a9fefda8bfae54549cb5e753909 removed the gallery images from this repo. We will help you get them into the app store format and will host them here: https://github.com/getumbrel/umbrel-apps-gallery
-  https://github.com/getumbrel/umbrel-apps/pull/649/commits/c4e471a707676812e47ba4303074d255f363526c https://github.com/getumbrel/umbrel-apps/pull/649/commits/5906a3e27e8f8e898e34b18f22cb91ec3deef30d fixed indentation in the umbrel-app.yml and added the missing `dependencies` field which was resulting in the app not showing up in the app store like you noted here: https://github.com/getumbrel/umbrel-apps/pull/649#issuecomment-1651009381
- https://github.com/getumbrel/umbrel-apps/pull/649/commits/6e496576d47371c0e7bb914f6ba26e4da0c2d457 added a [unique deterministic password](https://github.com/getumbrel/umbrel-apps/blob/master/README.md#:~:text=exposed%20at%0A%20%20%20%20%20%20%23-,%24APP_PASSWORD,-%2D%20Unique%20plain%20text) for the admin page so that users will be presented with an password in the app store that they can use to visit the admin page and set up their server
- https://github.com/getumbrel/umbrel-apps/pull/649/commits/5019b351278ba70e3467db6d8f13987003a61ede changed the port back to 7000, which is actually a free port right now that can be used on Umbrel. The port clash you were having early on was likely due to some testing setup (e.g., a rogue container still running on port 7000).
- https://github.com/getumbrel/umbrel-apps/pull/649/commits/0db05dec30b614325bef1f20d4e0bba8177fb2da https://github.com/getumbrel/umbrel-apps/pull/649/commits/02834df16836dc596b066491058a33404d801ded changed the bind mount slightly and committed the directory to this repo so that it gets set up with the proper permissions when a user installs the app.
- added a specific image tag and multi-arch digest so that we lock in the exact image we have tested.

> It might be in a good state as is after the changes suggested by @IMPranshu, but I am unable to test on my Pi. My home network has a reverse proxy forwarding to my various systems and services, and I can't seem to configure that with umbrel's authentication redirection

Ya, it can be tricky to set up a reverse proxy due to the our app proxy that sits in front of the app. My suggestion to make this as easy as possible for users is to have instructions in the app store description to download the C[loudflare Tunnel app from the app store](https://apps.umbrel.com/app/cloudflared) and follow instructions in the app to set up a Cloudflare Tunnel to the databag container running on their umbrel (so the databag_web_1 container) like this: 

image

This will bypass the app proxy, but your app has its own built-in auth. I have tested this earlier today and it works well. I was able to configure the server pretty easily by following a tutorial online but it may be worth linking something in the app description or adding additional instructions so that people realize you need to access the admin account first before setting up users. What do you think of this?

I will start the process for getting the gallery images created.

**Comment by nmfretz** [comment]
Hey @balzack, sorry again for the delay in reviewing on my end. I went through this today and pushed some commits to get it working on Umbrel. Everything is working well! Here are the changes I made:
- https://github.com/getumbrel/umbrel-apps/pull/649/commits/35a7a9b4fd576a9fefda8bfae54549cb5e753909 removed the gallery images from this repo. We will help you get them into the app store format and will host them here: https://github.com/getumbrel/umbrel-apps-gallery
-  https://github.com/getumbrel/umbrel-apps/pull/649/commits/c4e471a707676812e47ba4303074d255f363526c https://github.com/getumbrel/umbrel-apps/pull/649/commits/5906a3e27e8f8e898e34b18f22cb91ec3deef30d fixed indentation in the umbrel-app.yml and added the missing `dependencies` field which was resulting in the app not showing up in the app store like you noted here: https://github.com/getumbrel/umbrel-apps/pull/649#issuecomment-1651009381
- https://github.com/getumbrel/umbrel-apps/pull/649/commits/6e496576d47371c0e7bb914f6ba26e4da0c2d457 added a [unique deterministic password](https://github.com/getumbrel/umbrel-apps/blob/master/README.md#:~:text=exposed%20at%0A%20%20%20%20%20%20%23-,%24APP_PASSWORD,-%2D%20Unique%20plain%20text) for the admin page so that users will be presented with an password in the app store that they can use to visit the admin page and set up their server
- https://github.com/getumbrel/umbrel-apps/pull/649/commits/5019b351278ba70e3467db6d8f13987003a61ede changed the port back to 7000, which is actually a free port right now that can be used on Umbrel. The port clash you were having early on was likely due to some testing setup (e.g., a rogue container still running on port 7000).
- https://github.com/getumbrel/umbrel-apps/pull/649/commits/0db05dec30b614325bef1f20d4e0bba8177fb2da https://github.com/getumbrel/umbrel-apps/pull/649/commits/02834df16836dc596b066491058a33404d801ded changed the bind mount slightly and committed the directory to this repo so that it gets set up with the proper permissions when a user installs the app.
- added a specific image tag and multi-arch digest so that we lock in the exact image we have tested.

> It might be in a good state as is after the changes suggested by @IMPranshu, but I am unable to test on my Pi. My home network has a reverse proxy forwarding to my various systems and services, and I can't seem to configure that with umbrel's authentication redirection

Ya, it can be tricky to set up a reverse proxy due to the our app proxy that sits in front of the app. My suggestion to make this as easy as possible for users is to have instructions in the app store description to download the C[loudflare Tunnel app from the app store](https://apps.umbrel.com/app/cloudflared) and follow instructions in the app to set up a Cloudflare Tunnel to the databag container running on their umbrel (so the databag_web_1 container) like this: 

image

This will bypass the app proxy, but your app has its own built-in auth. I have tested this earlier today and it works well. I was able to configure the server pretty easily by following a tutorial online but it may be worth linking something in the app description or adding additional instructions so that people realize you need to access the admin account first before setting up users. What do you think of this?

I will start the process for getting the gallery images created.

**Comment by nmfretz** [comment]
Also, a couple considerations:
- If you'd like to disable the authentication that our app proxy provides you can add the env var `PROXY_AUTH_ADD: "false"` to the app_proxy container. Doing this may make your redirection issues go away.
- I see that the databag image size is currently around 1GB, which could mean that app install may take quite a while for users on slower internet connections. This isn't a blocker, but it would probably help at some point to reduce the image size.

**Comment by balzack** [comment]
Thank you for your help getting this to work within Umbrel!

your comments:

1. inking something in the app description or adding additional instructions so that people realize you need to access the admin account first before setting up users. What do you think of this?
- I think you are right; it's not clear how to obtain a user account. I will add some text in the login screen suggesting how to do this in the next release. Is your suggestion to update the .yml file with some instructions as well? 

2. disable the authentication that our app proxy provides
- I'll give this a try, seems like that should work

3. databag image size is currently around 1GB
- Thanks for calling this out. The image does contain a lot of dev tools that are aren't needed. I will remove this in the next release.

**Comment by nmfretz** [comment]
You're welcome! Glad I can help get this awesome submission to the finish line.

>  I think you are right; it's not clear how to obtain a user account. I will add some text in the login screen suggesting how to do this in the next release. Is your suggestion to update the .yml file with some instructions as well?

With text in the login screen, then I don't we need any admin account instructions in the .yml (under `description`). But if you want to send this to the app store prior to the next release then I think we should add something simple that tells users to log in to the admin account to set up the server.

Other than that we'd need either instructions or a link to instructions explaining how users can access the app with https.

> The image does contain a lot of dev tools that are aren't needed. I will remove this in the next release.

Excellent. Definitely not a deal breaker for this initial release if you want to go to the app store now.

**Comment by balzack** [comment]
I took a pass at the description following your suggestions; feel free to change it anyway you see fit.

The next release where I will trim the docker image and add the login text is at least a month away. I am currently reworking the browser app to support dark mode as well as provide multiple translations. I wouldn't mind releasing the current image, but if you prefer to wait for the next release, that's also fine.

Thanks for all your support!

**Comment by nmfretz** [comment]
Thanks for your patience on this @balzack as we hustle to get umbrelOS 1.0 out.

Databag is going live to the app store! Thanks again for all your hard work on this, we really appreciate it. 

FYI, we can change app description/instructions any time even without updating the app, just let me know whenever you need help.

---------------------------
I ended up tweaking the setup instructions a bit.
The gallery images and icon are stored here: https://github.com/getumbrel/umbrel-apps-gallery/tree/master/databag

image

**Pull request merged**